### PR TITLE
Fix [TF1,TF2,TF3] Save methods

### LINKED ISSUE #13927: `[TF1,TF2,TF3]::Save` works incorrectly when called with default arguments (all 0)
### Check duplicate issues.

- [X] Checked for duplicates

### Description

When TF1::Save with default arguments (all 0), stored buffer incorrect and cannot be used.
When all zeros are provided as argument, it suppose to write function value in pre-defined region between fXmin and fXmax. But makes failure by 0.5 bin. 

### Reproducer

Simple reproducer:
```
void test_save()
{
   auto f1 = new TF1("linear", "x", -5, 5);
   f1->SetNpx(10);
   f1->Draw("l");
   Double_t xx[1] = { 0 };
   // this works properly
   f1->Save(-5, 5, 0, 0, 0, 0);
   printf("Get saved at 0 %f\n", f1->GetSave(xx));
   // calling with default args make mistake by 0.5 bin
   f1->Save(0, 0, 0, 0, 0, 0);
   printf("Get saved at 0 %f\n", f1->GetSave(xx));
}
```
[test_save.cxx.txt](https://github.com/root-project/root/files/13117827/test_save.cxx.txt)

It expects to return 0 for x==0, but for second case it returns 0.5.
Running the example :

```
Get saved at 0 0.000000
Get saved at 0 0.500000
```

### ROOT version

master, all recent releases

### Installation method

build from source

### Operating system

OpenSUSE Linux

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by linev**
Since the beginning when 0 ranges were specified by Save method, 
half a bin was add to coordinate. But wrong number of bins was stored and 
therefore produced buffer was not able to use. Error was from very beginning  - 
already by initial import from rcs to svn 23 years ago.

Now proper number of bins are stored. It still stores middle of bins when `f1->Save(0,0,0,0,0,0)` is called.

There was also failure to use such saved buffer when trying to get approximation at upper boundaries. Now it is fixed.

Adding tests for all TF1/TF2/TF3.

Fixes #13927

### CODE REVIEW

**Review by pcanal — DISMISSED** [review]
Do we have a test for this broken case *and* the previously working cases?

Why was the .5 implemented originally?

**Comment by linev** [comment]
I mention already - this code was buggy from the beginning. 
But only if you call `f1->Save(0,0,0,0,0,0)`. Seems to be nobody was using it this way.

I am not sure if we have testing of `TF1::Save()` and `TF1::GetSave()` methods.

**Comment by pcanal** [comment]
> I am not sure if we have testing of TF1::Save() and TF1::GetSave() methods.

We need to add the test for both `f1->Save(0,0,0,0,0,0)` and the other cases.

**Comment by linev** [comment]
@pcanal 

I add testing for TF1/TF2/TF3 Save() / GetSave() methods.

File: hist/hist/test/test_tf2.cxx [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+TEST(TF2, Save)
+{
+   TF2 linear("linear", "x+y", -10, 10, -10, 10);
+   linear.SetNpx(20);
+   linear.SetNpy(20);
+
+   // this works properly since a while
+   linear.Save(-10, 10, -10, 10, 0, 0);
+   for (Double_t x = -10.; x < 10.; x += 0.33) {
+      for (Double_t y = -10.; y < 10.; y += 0.44) {
```
Comment (pcanal) [inline hist/hist/test/test_tf2.cxx:14]: Is there a few iteration that ends 'exactly' on the boundary?

Comment (linev) [reply hist/hist/test/test_tf2.cxx:14]: It is not important. One tests interpolation.

**Review by lmoneta — COMMENTED** [review]
Thanks Sergei for making now consistent.
However, I think it makes more sense to me to evaluate and save the function values on the bin centres and not on the edges. Was this behaviour changed in the past in the usage of Save ? 
Evaluating on bin centres makes it more consistent with what is done when 
drawing the function in `TF1::Paint` where it is used the underlined histogram.

**Comment by linev** [comment]
> I think it makes more sense to me to evaluate and save the function values on the bin centres and not on the edges.

It depends that one wants to do with these saved values. 
To use with `TF1::GetSave()` - then current approach is better. One covers complete range.
If values at bin centers will be save - than calculating of function at xmin/xmax will not be possible while extrapolation is not implemented in `GetSave()`.

For function painting bin centers is better. 
But this one can handle in TWebCanvas. 
There `f1->Save()` method called directly and one can shrink range by half of the bin.

Therefore I prefer to have reliable function which saves points exactly at the range which was specified.

**Comment by lmoneta** [comment]
I see your point, but we have to be careful in case the function has a singularity at the edge. 
Here is a problematic example which does not work correctly now:
```
auto f1 = new TF1("F1",[](double *x, double *){ return 1./x[0]; },0,1,0); 
f1->Draw();   // works fine (values are computed at bin centres)
TF1 * f2 = (TF1*)f1->Clone("f2");   // when cloning TF1::Save is used 
f2->Draw();
Warning in <TCanvas::ResizePad>: Inf/NaN propagated to the pad. Check drawn objects.
Warning in <TCanvas::ResizePad>: rf601_intminuit height changed from 0 to 10

Error in <TGaxis::PaintAxis>: wmin (0.000000) == wmax (0.000000)
```
This is because TF1::Save uses the bin boundary and not the centre

**Comment by linev** [comment]
> This is because TF1::Save uses the bin boundary and not the centre

Yes, I know. For instance in `tutorials/fit/fit1.C` same problem happens.

As I said - TWebCanvas can shrink saved range by half of the bin. 
And this I will be implementing now.

**Comment by linev** [comment]
I will redo this PR.

File: hist/hist/src/TF1.cxx [hunk 2]
```diff
@@ -3168,43 +3171,44 @@ void TF1::Save(Double_t xmin, Double_t xmax, Double_t, Double_t, Double_t, Doubl
          TH1 *h = (TH1 *)fParent;
          Int_t bin1 = h->GetXaxis()->FindBin(xmin);
          Int_t bin2 = h->GetXaxis()->FindBin(xmax);
-         int fNsave = bin2 - bin1 + 4;
-         //fSave  = new Double_t[fNsave];
-         fSave.resize(fNsave);
+         int nsave = bin2 - bin1 + 4;
```
Comment (lmoneta) [inline hist/hist/src/TF1.cxx:3174]: Why are you using now `nsave` and not `fNSave` ? I don't see where `fNSave` is set to the number of saved points

**Review by lmoneta — COMMENTED** [review]
Thank you for updating the PR. I have a comment on the variable `fNSave`, maybe it is not needed anymore, then we should remove from the data member and make an I/O rule for backward compatibility in I/O

File: hist/hist/src/TF1.cxx (hunk 2 above)
Comment (linev) [reply hist/hist/src/TF1.cxx:3174]: fNsave was data member some time ago. Now I rename it to avoid confusion.

Comment (lmoneta) [reply hist/hist/src/TF1.cxx:3174]: OK, thanks for clarifying this! 
It would be good to rename fSave to save also in the `TF1::GetSave` function

**Review by lmoneta — APPROVED** [review]
LGTM! 
Thank you Sergei for this fix!

**Pull request merged**